### PR TITLE
Last changes

### PR DESCRIPTION
**Pull request by schernolyas**
All my changes.
Merge and rebase with main 'hibernate/hibernate-orm' included

Please see to tests:
1) org.hibernate.datastore.ogm.orientdb.OrientDbAssociationTest
2) org.hibernate.datastore.ogm.orientdb.OrientDBSimpleTest

### CODE REVIEW

File: orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java [hunk 1]
```diff
@@ -0,0 +1,559 @@
...
+			}
+			Map<String, Object> dbValuesMap = entityQueries.get( key.getMetadata() ).findEntity( provider.getConnection(), key );
+			if ( dbValuesMap == null || ( dbValuesMap != null && dbValuesMap.isEmpty() ) ) {
+				return null;
+			}
+			return new Tuple(
+					new OrientDBTupleSnapshot( dbValuesMap, tupleContext.getAllAssociatedEntityKeyMetadata(), tupleContext.getAllRoles(), key.getMetadata() ) );
+		}
+		catch (SQLException e) {
+			log.error( "Can not find entity", e );
```
Comment (DavideD) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java]: What exception are you catching exactly?
An error in the SQL or a entity not found?

File: orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java [hunk 2]
```diff
@@ -0,0 +1,559 @@
...
+import org.hibernate.datastore.ogm.orientdb.utils.EntityKeyUtil;
+import org.hibernate.ogm.dialect.query.spi.QueryableGridDialect;
+import org.hibernate.ogm.model.key.spi.RowKey;
+import org.hibernate.service.spi.ServiceRegistryAwareService;
+import org.hibernate.service.spi.ServiceRegistryImplementor;
+
+/**
+ * @author Sergey Chernolyas (sergey.chernolyas@gmail.com)
+ */
+public class OrientDBDialect extends BaseGridDialect implements MultigetGridDialect, QueryableGridDialect<String>,
```
Comment (DavideD) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java]: Why did you add the MultigetGridDialect? We don't need it at the moment

File: orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java [hunk 3]
```diff
@@ -0,0 +1,559 @@
...
+		ServiceRegistryAwareService, SessionFactoryLifecycleAwareDialect, IdentityColumnAwareGridDialect {
```
Comment (DavideD) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java]: No need to implement IdentityColumnAwareGridDialect right now

File: orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java [hunk 4]
```diff
@@ -0,0 +1,559 @@
...
+	private Map<AssociationKeyMetadata, OrientDBAssociationQueries> associationQueries;
+	private Map<EntityKeyMetadata, OrientDBEntityQueries> entityQueries;
+
+	public OrientDBDialect(OrientDBDatastoreProvider provider) {
+		this.provider = provider;
+	}
+
+	@Override
+	public Tuple getTuple(EntityKey key, TupleContext tupleContext) {
+		log.info( "getTuple:EntityKey:" + key + "; tupleContext" + tupleContext + " tupleContext.getClass():" + tupleContext.getClass() );
```
Comment (DavideD) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java]: This kind of logging should be on trace or debug level,

Comment (DavideD) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java]: We also tends to avoid string concatenation for log messages for performance reasons.

File: orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java [hunk 5]
```diff
@@ -0,0 +1,559 @@
...
+
+		try {
+			if ( !entityQueries.containsKey( key.getMetadata() ) ) {
+				// throw new EntityNotFoundException("Key : "+key.toString());
+				return null;
+			}
```
Comment (DavideD) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java]: This case should never happen, since we have all the metafatas at startup

File: orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java [hunk 6]
```diff
@@ -0,0 +1,559 @@
...
+			Map<String, Object> dbValuesMap = entityQueries.get( key.getMetadata() ).findEntity( provider.getConnection(), key );
+			if ( dbValuesMap == null || ( dbValuesMap != null && dbValuesMap.isEmpty() ) ) {
+				return null;
```
Comment (DavideD) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java]: You can write this as:
`(dbValuesMap == null || dbValuesMap.isEmpty()  )`

File: orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java (hunk 1 above)
Comment (schernolyas) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java]: SQL exception.  About "entity not found" ... You are right. The code maust throw "entity not found"  exception. I will fix the incorrection

File: orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java (hunk 6 above)
Comment (DavideD) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java]: You should always be able to find the queries that you need to specific metadata. no need to check

File: orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java [hunk 7]
```diff
@@ -0,0 +1,559 @@
...
+	@Override
+	public void forEachTuple(ModelConsumer consumer, EntityKeyMetadata... entityKeyMetadatas) {
+		throw new UnsupportedOperationException( "forEachTuple!.Not supported yet." );
+	}
+
+	@Override
+	public List<Tuple> getTuples(EntityKey[] keys, TupleContext tupleContext) {
+		ArrayList<Tuple> tuples = new ArrayList<>( keys.length );
+		for ( EntityKey key : keys ) {
+			log.info( "getTuples:EntityKey:" + key + "; tupleContext" + tupleContext );
```
Comment (DavideD) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java]: you should use trace/debug fot this kind of logs

File: orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java (hunk 2 above)
Comment (schernolyas) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java]: I have implemented the interface.Why? ....I don't know  I got exception and impelemnt the method.  It is not error, isn't it?

File: orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java [hunk 8]
```diff
@@ -0,0 +1,559 @@
...
+
+		StringBuilder queryBuffer = new StringBuilder();
+		String dbKeyName = key.getColumnNames()[0];
+		Object dbKeyValue = key.getColumnValues()[0];
+
+		for ( int i = 0; i < key.getColumnNames().length; i++ ) {
+			String columnName = key.getColumnNames()[i];
+			Object columnValue = key.getColumnValues()[i];
+			log.info( "EntityKey: columnName: " + columnName + ";columnValue: " + columnValue + " (class:" + columnValue.getClass().getName() + ");" );
+		}
```
Comment (DavideD) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java]: Is this just for debugging?

File: orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java (hunk 3 above)
Comment (schernolyas) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java]: I have implement the interface too

File: orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java (hunk 8 above)
Comment (DavideD) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java]: It should not be included in the PR

File: orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java (hunk 4 above)
Comment (schernolyas) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java]: You right absolutly.. The log message is debug and analyse Hibernate code for. I will remove or change log level

File: orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java (hunk 1 above)
Comment (DavideD) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java]: But why is that an exception? Is it OrientDB? In general the id might not exists in the datastore.

It seems a valid use case

File: orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java (hunk 2 above)
Comment (DavideD) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java]: Yeah, ok, the implemntation is not correct though.
Could you remove the interface for now?
It should not give you any exception as this is an optional interface to add.
If there are exceptions the problem is somewhere else

File: orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java (hunk 3 above)
Comment (DavideD) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java]: Ah, ok. I thought you said that it wasn't working.
My bad

File: orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java [hunk 9]
```diff
@@ -0,0 +1,559 @@
...
+					EntityKeyUtil.setFieldValue( queryBuffer, tuple.get( columnName ) );
+					queryBuffer.append( "," );
+				}
+				queryBuffer.append( dbKeyName ).append( " = " );
+				EntityKeyUtil.setFieldValue( queryBuffer, dbKeyValue );
+			}
+
+			log.info( "insertOrUpdateTuple:Key:" + dbKeyName + " (" + dbKeyValue + ").  query:" + queryBuffer.toString() );
+			PreparedStatement pstmt = connection.prepareStatement( queryBuffer.toString() );
+			log.info( "insertOrUpdateTuple:Key:" + dbKeyName + " (" + dbKeyValue + "). inserted or updated: " + pstmt.executeUpdate() );
```
Comment (DavideD) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java]: Shouldn't the query already be in OrientDBEntityQueries?All the EnittyKeyMetadatas are generated at startup, you can create all the queries you need for CRUD operations and cache them.

File: orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java [hunk 10]
```diff
@@ -0,0 +1,559 @@
...
+		}
+		catch (SQLException e) {
+			log.error( "Can not find entity", e );
+			throw new RuntimeException( e );
```
Comment (DavideD) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java]: Not finding an entity should not be an exception as far as I understand (you are inserting a new one)
Or the SQLException refers to something else?

Also, have a look at the other dialects, for the way we generate and throws exceptions and feel free to ask questions if something is not clear (we generate all the classes for logging and exception during build time, so it might look a bit strange)

File: orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java [hunk 11]
```diff
@@ -0,0 +1,559 @@
...
+			insertQuery.append( " set " );
+		}
+
+		String dbKeyName = entityKeyMetadata.getColumnNames()[0];
+		Long dbKeyValue = null;
+		Connection connection = provider.getConnection();
+
+		if ( dbKeyName.equals( OrientDBConstant.SYSTEM_RID ) ) {
+			// use @RID for key
+			throw new UnsupportedOperationException( "Can not use @RID as primary key!" );
```
Comment (DavideD) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java]: This should be an `HibernateException`, check `org.hibernate.datastore.ogm.orientdb.logging.impl.Log` and create a new error log for it.

Comment (DavideD) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java]: (A nicer explanation on the why you cannot use @RID as primary key would help)

File: orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java (hunk 1 above)
Comment (schernolyas) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java]: Yes  I right absolutly. Thank You. I will fix it.

File: orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java (hunk 2 above)
Comment (schernolyas) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java]: Ok. I will remove it right now

File: orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java (hunk 3 above)
Comment (schernolyas) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java]: Need remove?

File: orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java (hunk 2 above)
Comment (schernolyas) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java]: Executed.

File: orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java [hunk 12]
```diff
@@ -0,0 +1,559 @@
...
+				String key = entry.getKey();
+				TypedGridValue value = entry.getValue();
+				log.info( "key: " + key + "; type:" + value.getType().getName() + "; value:" + value.getValue() );
+				// @todo move to Map
+				if ( value.getType().getName().equals( "string" ) ) {
+					pstmt.setString( 1, (String) value.getValue() );
+				}
+				if ( value.getType().getName().equals( "long" ) ) {
+					pstmt.setLong( 1, (Long) value.getValue() );
+				}
```
Comment (DavideD) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java]: Does this only work for Long values and String?
We should try to add all the other values
(I think it would be easier to treat this as a seprate issue, but it's your call)

File: bom/pom.xml [hunk 13]
```diff
@@ -27,6 +27,7 @@
         <infinispanVersion>8.0.2.Final</infinispanVersion>
         <mongodbVersion>3.0.3</mongodbVersion>
         <fongodbVersion>2.0.2</fongodbVersion>
+        <orientdbVersion>[2.1.9,)</orientdbVersion>
```
Comment (DavideD) [inline bom/pom.xml]: We don't use range for versions, it migh tlead to some unpredictable results.
Please, change it to a fixed version.

File: orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBProperties.java [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.datastore.ogm.orientdb;
+
+import org.hibernate.ogm.cfg.OgmProperties;
+
+/**
+ * @author Sergey Chernolyas (sergey.chernolyas@gmail.com)
+ */
+public class OrientDBProperties implements OgmProperties {
```
Comment (DavideD) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBProperties.java]: Maybe we can wait addint this file when we have a use for it

File: orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBProperties.java [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.datastore.ogm.orientdb;
+
+import org.hibernate.ogm.cfg.OgmProperties;
+
+/**
+ * @author Sergey Chernolyas (sergey.chernolyas@gmail.com)
```
Comment (DavideD) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBProperties.java]: Could you add your name in the author without the email?
In past we removed it almost everywhere because it causes some issues when generating the JavaDoc

File: orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/constant/OrientDBConstant.java [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+ * @author Sergey Chernolyas <sergey.chernolyas@gmail.com>
+ */
+
+public class OrientDBConstant {
+
+	public static final String SYSTEM_VERSION = "@version";
+	public static final String SYSTEM_RID = "@rid";
+
+	public static final Set<String> SYSTEM_FIELDS;
+        public static final Set<String> LINK_FIELDS;
```
Comment (DavideD) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/constant/OrientDBConstant.java]: THe formattation of the code is strange.
It would also be usefull to add some comments to describe this constants

File: orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java [hunk 17]
```diff
@@ -0,0 +1,559 @@
...
+		return queryMap;
+	}
+
+	@Override
+	public GridType overrideType(Type type) {
+		log.info( "overrideType:" + type.getName() + ";" + type.getReturnedClass() );
+		GridType gridType = null;
+		if ( type.getName().equals( "com.orientechnologies.orient.core.id.ORecordId" ) ) {
+			gridType = ORecordIdGridType.INSTANCE;
+		} else if ( type.getName().equals( "com.orientechnologies.orient.core.db.record.ridbag.ORidBag" ) ) {
```
Comment (DavideD) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java]: You should check the type using `instanceof`.
Someone could extends the class or could have a class with the same package and name.

File: orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java [hunk 18]
```diff
@@ -0,0 +1,559 @@
...
+			gridType = ORidBagGridType.INSTANCE;
+		}
+		else {
+			gridType = super.overrideType( type ); // To change body of generated methods, choose Tools | Templates.
```
Comment (DavideD) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java]: If it's possible, it would be nice if you could send PRs without this kind of comments.

File: orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java (hunk 17 above)
Comment (DavideD) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java]: So, are these the only types you need to check?
Does OrientDB support all the others? Have checked `BuiltinTypeTest`?

File: orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/dialect/impl/OrientDBAssociationSnapshot.java [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+import java.util.Set;
+import org.hibernate.datastore.ogm.orientdb.logging.impl.Log;
+import org.hibernate.datastore.ogm.orientdb.logging.impl.LoggerFactory;
+import org.hibernate.ogm.model.key.spi.RowKey;
+import org.hibernate.ogm.model.spi.AssociationSnapshot;
+import org.hibernate.ogm.model.spi.Tuple;
+
+/**
+ * @author Sergey Chernolyas (sergey.chernolyas@gmail.com)
+ * @see org.hibernate.ogm.datastore.neo4j.dialect.impl.Neo4jAssociationSnapshot
```
Comment (DavideD) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/dialect/impl/OrientDBAssociationSnapshot.java]: Remember to update the comments whe you copy fomr existing classes

File: orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/dialect/impl/OrientDBEntityQueries.java [hunk 20]
```diff
@@ -0,0 +1,209 @@
...
+			}
+			else {
+				// search by business key
+				query.append( entityKey.getTable() ).append( " WHERE " ).append( dbKeyName ).append( " = " );
+				EntityKeyUtil.setFieldValue( query, dbKeyValue );
+			}
+		}
+		else {
+			// query.append(entityKeyMetadata.getTable());
+			throw new UnsupportedOperationException( "Not supported yet." );
```
Comment (DavideD) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/dialect/impl/OrientDBEntityQueries.java]: It would be nice if you could add a more descriptive message, what's not supported?

File: orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java (hunk 3 above)
Comment (DavideD) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java]: Let's keep the interface `IdentityColumnAwareGridDialect`

File: orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/query/impl/OrientDBQueryParser.java [hunk 21]
```diff
@@ -0,0 +1,145 @@
...
+import static org.parboiled.BaseParser.EOI;
+import org.parboiled.Rule;
+import org.parboiled.annotations.SkipNode;
+import org.parboiled.annotations.SuppressSubnodes;
+import org.parboiled.support.StringVar;
+
+/**
+ * @author Sergey Chernolyas (sergey.chernolyas@gmail.com)
+ */
+public class OrientDBQueryParser extends BaseParser<ParameterParser.Recognizer> {
```
Comment (DavideD) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/query/impl/OrientDBQueryParser.java]: Let's keep these classes for another time

File: orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/utils/TupleUtil.java [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+			for ( int i = 0; i < metadata.getColumnCount(); i++ ) {
+				int fn = i + 1;
+				classPropertyNames.add( metadata.getColumnName( fn ) );
+			}
+		}
+
+		return classPropertyNames;
+	}
+@Deprecated
+	public static Set<String> getFieldsForJoin(TupleContext tupleContext, Collection<String> classPropertyNames) {
```
Comment (DavideD) [inline orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/utils/TupleUtil.java]: Strange indentation in this class,
why are you adding a deprecated method?

File: orientdb/src/main/java/org/hibernate/search/orientdb/bridge/ORecordIdTwoWayStringBridge.java [hunk 23]
```diff
@@ -0,0 +1,28 @@
...
+ */
+package org.hibernate.search.orientdb.bridge;
+
+import com.orientechnologies.orient.core.id.ORecordId;
+import org.hibernate.search.bridge.TwoWayStringBridge;
+
+/**
+ * @author Sergey Chernolyas (sergey.chernolyas@gmail.com)
+ */
+public class ORecordIdTwoWayStringBridge implements TwoWayStringBridge {
```
Comment (DavideD) [inline orientdb/src/main/java/org/hibernate/search/orientdb/bridge/ORecordIdTwoWayStringBridge.java]: I don't think you need this class, OGM does not deal with StringBridges (This is Hibernate Search stuff)

File: orientdb/src/test/java/org/hibernate/datastore/ogm/orientdb/OrientDBSimpleTest.java [hunk 24]
```diff
@@ -0,0 +1,280 @@
...
+import static org.junit.Assert.*;
+import org.junit.FixMethodOrder;
+import org.junit.runners.MethodSorters;
+
+/**
+ * Test checks CRUD for simple entities (without links with other entities)
+ * 
+ * @author Sergey Chernolyas (sergey.chernolyas@gmail.com)
+ */
+@FixMethodOrder(MethodSorters.NAME_ASCENDING)
```
Comment (DavideD) [inline orientdb/src/test/java/org/hibernate/datastore/ogm/orientdb/OrientDBSimpleTest.java]: Why do you need this? It smells :)

File: orientdb/src/test/java/org/hibernate/datastore/ogm/orientdb/OrientDBSimpleTest.java [hunk 25]
```diff
@@ -0,0 +1,280 @@
...
+
+	@AfterClass
+	public static void tearDownClass() {
+		LOG.log( Level.INFO, "start" );
+		if ( em != null ) {
+			em.close();
+			emf.close();
+
+		}
+		MemoryDBUtil.getOrientGraphFactory().close();
```
Comment (DavideD) [inline orientdb/src/test/java/org/hibernate/datastore/ogm/orientdb/OrientDBSimpleTest.java]: Shouldn't this be in a finally block (jus in case)

Comment (DavideD) [inline orientdb/src/test/java/org/hibernate/datastore/ogm/orientdb/OrientDBSimpleTest.java]: By the way, if your test class extends `OgmTestCase`  and the datastore provider is implemented correctly, the SessionFactory is created automagically and the database started and closed.

Comment (DavideD) [inline orientdb/src/test/java/org/hibernate/datastore/ogm/orientdb/OrientDBSimpleTest.java]: Have a look at the other tests implemnting `OgmTestCase`

File: orientdb/src/test/java/org/hibernate/datastore/ogm/orientdb/OrientDBSimpleTest.java [hunk 26]
```diff
@@ -0,0 +1,280 @@
...
+		}
+	}
+
+	@Test
+	public void test2FindPizza() {
+		LOG.log( Level.INFO, "start" );
+		try {
+			em.getTransaction().begin();
+			Pizza pizza = em.find( Pizza.class, Long.valueOf( 2L ) );
+			em.refresh( pizza );
```
Comment (DavideD) [inline orientdb/src/test/java/org/hibernate/datastore/ogm/orientdb/OrientDBSimpleTest.java]: Why the `refresh`?

File: orientdb/src/test/java/org/hibernate/datastore/ogm/orientdb/OrientDBSimpleTest.java [hunk 27]
```diff
@@ -0,0 +1,280 @@
...
+		}
+	}
+
+	@Test
+	public void test2FindCustomer() {
+		LOG.log( Level.INFO, "start" );
+		try {
+			em.getTransaction().begin();
+			Customer customer = em.find( Customer.class, Long.valueOf( 2L ) );
+			em.refresh( customer );
```
Comment (DavideD) [inline orientdb/src/test/java/org/hibernate/datastore/ogm/orientdb/OrientDBSimpleTest.java]: Why the `refresh`?

File: orientdb/src/test/java/org/hibernate/datastore/ogm/orientdb/OrientDBSimpleTest.java [hunk 28]
```diff
@@ -0,0 +1,280 @@
...
+			LOG.log( Level.INFO, "read entity properties:" );
+			LOG.log( Level.INFO, "customer.getbKey():{0}", customer.getbKey() );
+			LOG.log( Level.INFO, "customer.getName(): {0}", customer.getName() );
+			LOG.log( Level.INFO, "customer.getRid(): {0}", customer.getRid() );
```
Comment (DavideD) [inline orientdb/src/test/java/org/hibernate/datastore/ogm/orientdb/OrientDBSimpleTest.java]: Why all this log?
I think it would help if you have a look at the existing tests we have in core or in the other dialects

File: orientdb/src/test/java/org/hibernate/datastore/ogm/orientdb/OrientDBSimpleTest.java [hunk 29]
```diff
@@ -0,0 +1,280 @@
...
+	public void test4UpdateCustomer() {
+		LOG.log( Level.INFO, "start" );
+		try {
+			em.getTransaction().begin();
+			Long id = Long.valueOf( 2L );
+			Customer customer = em.find( Customer.class, id );
+			customer.setName( "Ivahoe" );
+			int oldVersion = customer.getVersion();
+			LOG.log( Level.INFO, "old version:{0}", oldVersion );
+			em.merge( customer );
```
Comment (DavideD) [inline orientdb/src/test/java/org/hibernate/datastore/ogm/orientdb/OrientDBSimpleTest.java]: You don't need to use merge in this case.

Just for reference, the em.merge() has a returned value not sure if it works as expected otherwise.

File: orientdb/src/test/java/org/hibernate/datastore/ogm/orientdb/OrientDBSimpleTest.java [hunk 30]
```diff
@@ -0,0 +1,280 @@
...
+	@Test
+	public void test5RefreshCustomer() {
+		LOG.log( Level.INFO, "start" );
+		try {
+			em.getTransaction().begin();
+			Long id = Long.valueOf( 2L );
+			Customer customer = em.find( Customer.class, id );
+			LOG.log( Level.INFO, "old rid:{0}", customer.getRid() );
+			ORecordId oldRid = customer.getRid();
+			em.refresh( customer );
```
Comment (DavideD) [inline orientdb/src/test/java/org/hibernate/datastore/ogm/orientdb/OrientDBSimpleTest.java]: THis is not needed

File: orientdb/src/test/java/org/hibernate/datastore/ogm/orientdb/jpa/Customer.java [hunk 31]
```diff
@@ -0,0 +1,120 @@
...
+
+/**
+ * @author chernolyassv
+ */
+@Entity
+@Indexed(index = "Customer")
+@NamedQueries({
+		@NamedQuery(name = "Customer.findAll", query = "SELECT c FROM Customer c"),
+		@NamedQuery(name = "Country.findByName", query = "SELECT c FROM Customer c WHERE c.name = :name") })
+public class Customer {
```
Comment (DavideD) [inline orientdb/src/test/java/org/hibernate/datastore/ogm/orientdb/jpa/Customer.java]: Let's remove the named querie

File: orientdb/src/test/java/org/hibernate/datastore/ogm/orientdb/jpa/package-info.java [hunk 32]
```diff
@@ -0,0 +1,19 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301  USA
+ */
+package org.hibernate.datastore.ogm.orientdb.jpa;
```
Comment (DavideD) [inline orientdb/src/test/java/org/hibernate/datastore/ogm/orientdb/jpa/package-info.java]: Wrong copyright file

File: orientdb/src/test/java/org/hibernate/datastore/ogm/orientdb/util/MemoryDBUtil.java [hunk 33]
```diff
@@ -0,0 +1,131 @@
...
+/**
+ * @author Sergey Chernolyas <sergey.chernolyas@gmail.com>
+ */
+public class MemoryDBUtil {
+
+	private static final Logger LOG = Logger.getLogger( MemoryDBUtil.class.getName() );
+	private static OrientGraphFactory factory;
+
+	@Deprecated
+	private static void addCommonProperties(OrientGraphNoTx graph, OrientVertexType type) {
```
Comment (DavideD) [inline orientdb/src/test/java/org/hibernate/datastore/ogm/orientdb/util/MemoryDBUtil.java]: Why `@Deprecated`?
This part has not been included yet. You can remove it or change it the way you prefer

File: orientdb/src/test/java/org/hibernate/datastore/ogm/orientdb/utils/OrientDBTestHelper.java [hunk 34]
```diff
@@ -0,0 +1,119 @@
...
+	}
+
+	@Override
+	public Map<String, String> getEnvironmentProperties() {
+		System.out.println( "org.hibernate.datastore.ogm.orientdb.util.OrientDBTestHelper.getEnvironmentProperties()" );
+		Map<String, String> properties = new HashMap<String, String>();
+		properties.put( "javax.persistence.jdbc.url", JDBC_URL );
+		properties.put( "javax.persistence.jdbc.driver", "com.orientechnologies.orient.jdbc.OrientJdbcDriver" );
+		properties.put( "javax.persistence.jdbc.user", "admin" );
+		properties.put( "javax.persistence.jdbc.password", "admin" );
```
Comment (DavideD) [inline orientdb/src/test/java/org/hibernate/datastore/ogm/orientdb/utils/OrientDBTestHelper.java]: In the end Hibernate OGM does not use the JDBC driver like Hibernate (in general).
OgmProperties should contains the one to set for OGM.
For this case in particolar, being embedded, the only properties that the user should be force to set is the dialect.

Check the other dialect examples.
